### PR TITLE
feature(new integration): onboarding mailmodo

### PR DESCRIPTION
**Pull request by rohithkaza**
## Description of the change

> Onboarding Mailmodo cloud mode destination

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/mailmodo/transform.js [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+  let { listName } = Config;
+
+  const payload = constructPayload(
+    message,
+    mappingConfig[ConfigCategory.IDENTIFY.name]
+  );
+
+  if (!payload.email) {
+    throw new CustomError("Email is required for identify call", 400);
+  }
```
Comment (ItsSudip) [inline v0/destinations/mailmodo/transform.js:53]: No need to check it here you can make it required: true in identify config.json

File: v0/destinations/mailmodo/transform.js [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+  if (isDefinedAndNotNullAndNotEmpty(message.address)) {
+    const { address1, address2 } = deduceAddressFields(message);
+    payload.data.address1 = address1;
+    payload.data.address2 = address2;
+  }
+  if (typeof listName === "string" && listName.trim().length === 0) {
+    listName = "rudderstack";
```
Comment (ItsSudip) [inline v0/destinations/mailmodo/transform.js:60]:
```suggestion
    listName = "Rudderstack";
```

File: v0/destinations/mailmodo/transform.js [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+  }
+  const valuesArray = [];
+  valuesArray.push(payload);
+  const resp = {};
+  resp.values = valuesArray;
+  resp.listName = listName;
```
Comment (ItsSudip) [inline v0/destinations/mailmodo/transform.js:66]:
```suggestion
  const valuesArray = [payload];
  const resp = { values: valuesArray, listName };
```

File: v0/destinations/mailmodo/transform.js [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+  const response = defaultRequestConfig();
+  if (resp) {
+    const responseBody = "JSON";
+    response.endpoint = `${BASE_URL}${endpoint}`;
+    response.headers = {
+      mmApiKey: `${apiKey}`,
+      "Content-Type": "application/json"
+    };
+    response.method = defaultPostRequestConfig.requestMethod;
+    response.body[`${responseBody}`] = removeUndefinedAndNullValues(resp);
```
Comment (ItsSudip) [inline v0/destinations/mailmodo/transform.js:37]:
```suggestion
    response.body.JSON= removeUndefinedAndNullValues(resp);
```

File: v0/destinations/mailmodo/transform.js [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+} = require("../../util");
+const { TRANSFORMER_METRIC } = require("../../util/constant");
+const { deduceAddressFields } = require("./utils");
+
+const responseBuilder = responseConfgs => {
+  const { resp, apiKey, endpoint } = responseConfgs;
+
+  const response = defaultRequestConfig();
+  if (resp) {
+    const responseBody = "JSON";
```
Comment (ItsSudip) [inline v0/destinations/mailmodo/transform.js:30]:
```suggestion
```

File: v0/destinations/mailmodo/transform.js [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+          // already transformed event
+          getEventChunks(event, identifyEventChunks, eventResponseList);
+          // slice according to batch size
+          if (
+            identifyEventChunks.length &&
+            (identifyEventChunks.length >= IDENTIFY_MAX_BATCH_SIZE ||
+              index === inputs.length - 1)
+          ) {
+            identifyArrayChunks.push(identifyEventChunks);
+            identifyEventChunks = [];
```
Comment (ItsSudip) [inline v0/destinations/mailmodo/transform.js:234]: This logic is hard to understand. Let's don't divide in chunks here. Put everything in array and break them in chunks later part.

File: v0/destinations/mailmodo/transform.js [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+    // listName will be "rudderstack", if it is not provided
+    if (typeof listName === "string" && listName.trim().length === 0) {
+      listName = "rudderstack";
+    }
+
+    let batchEventResponse = defaultBatchRequestConfig();
+
+    // Batch event into dest batch structure
+    chunk.forEach(ev => {
+      values.push(ev?.message?.body?.JSON?.values[0]);
```
Comment (ItsSudip) [inline v0/destinations/mailmodo/transform.js:148]:
```suggestion
      values.push(event?.message?.body?.JSON?.values[0]);
```

**Review by ItsSudip — CHANGES_REQUESTED** [review]

File: v0/destinations/mailmodo/data/identifyConfig.json [hunk 8]
```diff
@@ -0,0 +1,146 @@
+[
+  {
+    "destKey": "email",
+    "sourceKeys": ["traits.email", "context.traits.email", "properties.email"],
+    "sourceFromGenericMap": false,
```
[the remark this answers is not in the document]
Comment (sauravlal-rudderstack) [reply v0/destinations/mailmodo/data/identifyConfig.json:5]: `email` is present in genericfieldmapping we can use that.

File: v0/destinations/mailmodo/data/identifyConfig.json [hunk 9]
```diff
@@ -0,0 +1,123 @@
+[
+  {
+    "destKey": "email",
+    "sourceKeys": ["traits.email", "context.traits.email", "properties.email"],
+    "required": true
```
Comment (sauravlal-rudderstack) [inline v0/destinations/mailmodo/data/identifyConfig.json:5]: `email` is present in genericfieldmapping, we can use that.

File: v0/destinations/mailmodo/data/identifyConfig.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (rohithkaza) [reply v0/destinations/mailmodo/data/identifyConfig.json]: @ItsSudip 
Removed unnecessary 
`sourceFromGenericMap: false`

File: v0/destinations/mailmodo/data/identifyConfig.json [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+      "traits.zipcode",
+      "traits.address.zipcode",
+      "traits.address.postalcode",
+      "traits.address.postal_code",
+      "context.traits.zip",
+      "context.traits.zipcode",
+      "context.traits.address.zipcode",
+      "context.traits.address.postalcode",
+      "context.traits.address.postal_code"
+    ],
```
[the remark this answers is not in the document]
Comment (rohithkaza) [reply v0/destinations/mailmodo/data/identifyConfig.json]: Added few things to zipcode in genericfieldmapping

File: v0/destinations/mailmodo/data/identifyConfig.json [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+    "sourceFromGenericMap": false,
+    "required": false
+  },
+  {
+    "destKey": "data.anniversary_date",
+    "sourceKeys": [
+      "traits.anniversary_date",
+      "context.traits.anniversary_date",
+      "traits.anniversaryDate",
+      "context.traits.anniversaryDate"
```
[the remark this answers is not in the document]
Comment (rohithkaza) [reply v0/destinations/mailmodo/data/identifyConfig.json]: Done

File: v0/destinations/mailmodo/transform.js (hunk 1 above)
Comment (rohithkaza) [reply v0/destinations/mailmodo/transform.js]: Removed

File: v0/destinations/mailmodo/transform.js (hunk 2 above)
Comment (rohithkaza) [reply v0/destinations/mailmodo/transform.js]: Done

File: v0/destinations/mailmodo/transform.js (hunk 3 above)
Comment (rohithkaza) [reply v0/destinations/mailmodo/transform.js]: Done

File: v0/destinations/mailmodo/transform.js (hunk 4 above)
Comment (rohithkaza) [reply v0/destinations/mailmodo/transform.js]: Done

File: v0/destinations/mailmodo/transform.js (hunk 5 above)
Comment (rohithkaza) [reply v0/destinations/mailmodo/transform.js]: Done

File: v0/destinations/mailmodo/transform.js (hunk 6 above)
Comment (rohithkaza) [reply v0/destinations/mailmodo/transform.js]: Modified accordingly

File: v0/destinations/mailmodo/transform.js (hunk 7 above)
Comment (rohithkaza) [reply v0/destinations/mailmodo/transform.js]: Done

File: v0/destinations/mailmodo/transform.js [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+    // extracting destination
+    // from the first event in a batch
+    const { destination } = chunk[0];
+    const { apiKey } = destination.Config;
+    let { listName } = destination.Config;
+
+    // listName will be "Rudderstack", if it is not provided
+    if (typeof listName === "string" && listName.trim().length === 0) {
+      listName = "Rudderstack";
+    }
```
Comment (ItsSudip) [inline v0/destinations/mailmodo/transform.js:137]: This is not required as we are already setting `listName` in `process` function.

File: v0/destinations/mailmodo/transform.js [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+
+const identifyResponseBuilder = (message, { Config }) => {
+  const { apiKey } = Config;
+  let { listName } = Config;
+
+  const payload = constructPayload(
+    message,
+    mappingConfig[ConfigCategory.IDENTIFY.name]
+  );
+
```
[the remark this answers is not in the document]
Comment (rohithkaza) [reply v0/destinations/mailmodo/transform.js:49]: Added the support

File: v0/util/data/GenericFieldMapping.json [hunk 14]
```diff
@@ -81,9 +80,9 @@
     "context.traits.DOB"
   ],
 
-  "state":["traits.state", "context.traits.state"],
-  "country":["traits.country", "context.traits.country"],
-  "city":"traits.address.city",
+  "state": ["traits.state", "context.traits.state"],
+  "country": ["traits.country", "context.traits.country"],
+  "city": "traits.address.city",
```
[the remark this answers is not in the document]
Comment (rohithkaza) [reply v0/util/data/GenericFieldMapping.json]: Updated this one, though we are not using genericfieldmapping for `city`

File: v0/destinations/mailmodo/transform.js [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+          TRANSFORMER_METRIC.TRANSFORMER_STAGE.TRANSFORM
+        );
+        errorRespList.push(
+          getErrorRespEvents(
+            [event.metadata],
+            error.response ? error.response.status : 400,
+            error.message || "Error occurred while processing payload.",
+            errObj.statTags
+          )
+        );
```
[the remark this answers is not in the document]
Comment (rohithkaza) [reply v0/destinations/mailmodo/transform.js]: Yes, modified

File: v0/destinations/mailmodo/transform.js (hunk 12 above)
Comment (rohithkaza) [reply v0/destinations/mailmodo/transform.js]: Removed

**Review by ItsSudip — APPROVED** [review]

**Review by sauravlal-rudderstack — APPROVED** [review]

**Pull request merged**